File: backend/app/routers/forecasts.py
"""
Router for ML Price Trend Forecasts and Confidence Engine.
"""
from datetime import datetime
from typing import List, Literal
from fastapi import APIRouter, Query

from backend.app.schemas import ForecastSummary, RouteForecast, ConfidenceDistribution

router = APIRouter(prefix="/api/v1/forecasts", tags=["Forecasts"])


@router.get("/summary", response_model=ForecastSummary)
def get_forecast_summary(
    horizon: int = Query(7, description="Forecast horizon in days (7, 14, 30)")
):
    """Returns macroeconomic forecast summary across specified horizon."""
    h = horizon if horizon in (7, 14, 30) else 7
    movement_map = {7: 3.8, 14: 6.2, 30: 9.4}
    conf_map = {7: 94.2, 14: 88.5, 30: 81.0}

    movement = movement_map[h]
    current_idx = 124.60
    fc_idx = round(current_idx * (1 + movement / 100.0), 2)

    return {
        "horizonDays": h,
        "currentIndex": current_idx,
        "forecastIndex": fc_idx,
        "expectedMovementPct": movement,
        "confidencePct": conf_map[h],
        "signal": "rising",
        "generatedAt": datetime.utcnow().isoformat(),
        "modelVersion": "AIRIS-Forecaster-v2.4-Ensemble"
    }


@router.get("/routes", response_model=List[RouteForecast])
def get_route_forecasts(
    horizon: int = Query(7, description="Forecast horizon in days (7, 14, 30)")
):
    """Returns route-level 7/14/30-day forecast trajectories."""
    return [
        {"routeId": "DEL-BOM", "routeLabel": "DEL → BOM", "expectedChangePct": 4.5, "confidencePct": 94.2, "signal": "rising"},
        {"routeId": "BLR-DEL", "routeLabel": "BLR → DEL", "expectedChangePct": 3.8, "confidencePct": 92.0, "signal": "rising"},
        {"routeId": "BOM-GOI", "routeLabel": "BOM → GOI", "expectedChangePct": 11.4, "confidencePct": 89.5, "signal": "rising"},
        {"routeId": "DEL-BLR", "routeLabel": "DEL → BLR", "expectedChangePct": 2.2, "confidencePct": 91.8, "signal": "rising"},
        {"routeId": "BOM-BLR", "routeLabel": "BOM → BLR", "expectedChangePct": -0.8, "confidencePct": 88.0, "signal": "stable"},
        {"routeId": "DEL-CCU", "routeLabel": "DEL → CCU", "expectedChangePct": 1.4, "confidencePct": 86.5, "signal": "stable"},
        {"routeId": "DEL-HYD", "routeLabel": "DEL → HYD", "expectedChangePct": 2.9, "confidencePct": 90.1, "signal": "rising"},
        {"routeId": "BOM-HYD", "routeLabel": "BOM → HYD", "expectedChangePct": -1.2, "confidencePct": 87.4, "signal": "stable"}
    ]


@router.get("/confidence", response_model=List[ConfidenceDistribution])
def get_confidence_distribution():
    """Returns distribution of model confidence across all monitored routes."""
    return [
        {"level": "High", "routeCount": 32, "description": ">90% Model Confidence (Trunk corridors with deep liquidity)"},
        {"level": "Medium", "routeCount": 12, "description": "80% - 90% Confidence (Regional routes with seasonal volatility)"},
        {"level": "Low", "routeCount": 4, "description": "<80% Confidence (Thin routes with single-carrier dominance)"}
    ]

File: src/lib/api/forecasts.ts
import type { ConfidenceDistribution, ForecastHorizon, ForecastSummary, RouteForecast } from "@/types";
import { apiConfig, mockLatency, request } from "./client";
import * as mock from "@/lib/mock/handlers";

/**
 * Forecast domain — horizon summaries, per-route forecasts and confidence.
 *
 * REST mapping (FastAPI):
 *   GET /api/v1/forecasts/summary?horizon=7
 *   GET /api/v1/forecasts/routes?horizon=7
 *   GET /api/v1/forecasts/confidence
 */
export const forecastsApi = {
  getSummary(horizon: ForecastHorizon): Promise<ForecastSummary> {
    if (apiConfig.mockMode) return mockLatency(mock.mockGetForecastSummary(horizon));
    return request<ForecastSummary>(`/api/v1/forecasts/summary?horizon=${horizon}`);
  },

  getRouteForecasts(): Promise<RouteForecast[]> {
    if (apiConfig.mockMode) return mockLatency(mock.mockGetRouteForecasts());
    return request<RouteForecast[]>("/api/v1/forecasts/routes");
  },

  getConfidenceDistribution(): Promise<ConfidenceDistribution[]> {
    if (apiConfig.mockMode) return mockLatency(mock.CONFIDENCE_DISTRIBUTION);
    return request<ConfidenceDistribution[]>("/api/v1/forecasts/confidence");
  },
};
